### PR TITLE
Reports UI

### LINKED ISSUE #568: Create the reports view scaffolding
## What needs to be done?
- [x] Create the basic view, which should contain a blank UICollectionView/Recycler view and a date range button which should render the current date range

### Dependencies
Depends on #554

### LINKED ISSUE #570: Create the reports cell view model
## What needs to be done?
- [x] This should be an sealed class like below. Only add the cases, don't bother with the wrapped values at this point.
``` Kotlin
sealed class ReportsCellViewModel(
    class Total(val total: String) : ReportsCellViewModel()
    class billable(BillableCellData) : ReportsCellViewModel()
    class amounts([AmountsCellData] : ReportsCellViewModel()
    class BarChart(BarChartInfo) : ReportsCellViewModel()
    class DonutChart(DonutChartInfo) : ReportsCellViewModel()
    class DonutChartLegend(DonutChartLegendInfo) : ReportsCellViewModel()
    object Error : ReportsCellViewModel()
    object AdvancedReportsOnWeb : ReportsCellViewModel()
)
```
- [x] Configure the UICollectionView/Recycler view to receive cells of this type

### Dependencies
Depends on #554 and #568

### LINKED ISSUE #571: Create the error cell
## What needs to be done?
- [x] Create a error selector in the reports feature. This selector should check whether the reports totals and summary loadable are set to `.failed` and if they are, return a `ReportsCellViewModel.Error`.
- [x] Create the actual cell/view, and populate it with the appropriate data. Designs are are still pending for both platforms.
- [x] Discuss possible value animation with the team in #mobile-team

### Dependencies
Depends on #570

### LINKED ISSUE #572: Create the total hours reports cell
## What needs to be done?
- [x] Create a total hours selector in the reports feature. This selector should take the report totals from the state and return a `ReportsCellViewModel.Total` wrapping the a formatted string of the time interval from the report totals data class.
- [x] Create the actual cell/view, and populate it with the appropriate data. Designs are [here for iOS](https://toggl.invisionapp.com/d/main#/console/19800511/418226532/preview#project_console) and are still pending for Droid.
- [x] Discuss possible value animation with the team in #mobile-team

### Dependencies
Depends on #570

### LINKED ISSUE #573: Create the billable reports cell
## What needs to be done?
- [x] Create a `BillableCellData` data class like this:
``` Kotlin
data class BillableCellData(
    val time: String,
    val percentage: Double
)
```
- [x] Create a billable selector in the reports feature. This selector should take the report totals from the state and return a `ReportsCellViewModel.Billable` wrapping the billable time as a string and percentage from the report totals data class into a `BillableCellData`.
- [x] Create the actual cell/view, and populate it with the appropriate data. Designs are [here for iOS](https://toggl.invisionapp.com/d/main#/console/19800511/418226532/preview#project_console) and are still pending for Droid.
- [x] Discuss possible value animation with the team in #mobile-team

### Dependencies
Depends on #570

### LINKED ISSUE #574: Create the amounts reports cell
## What needs to be done?
- [x] Create a `AmountsCellData` data class like this:
``` Kotlin
data class AmountsCellData(
    val amount: Double,
    val currency: String
)
```
- [x] Create an amounts selector in the reports feature. This selector should take the report totals from the state and return a `ReportsCellViewModel.Amounts` wrapping an array of `AmountsCellData` that we can get by filtering the reports totals.
- [x] Create the actual cell/view, and populate it with the appropriate data. Designs are [here for iOS](https://toggl.invisionapp.com/d/main#/console/19800511/418226532/preview#project_console) and are still pending for Droid.
- [x] Discuss possible value animation with the team in #mobile-team

### Dependencies
Depends on #570

### LINKED ISSUE #577: Create the donut chart legend reports cell
## What needs to be done?
- [x] Create a `DonutChartLegendInfo` data class like this:
``` Kotlin
data class DonutChartLegendInfo(
    val title: String,
    val color: String,
    val value: String,
    val percentage: Double
)
```
- [x] Create an donut chart legends selector in the reports feature. This selector should take the report summary from the state and return an array of `ReportsCellViewModel.DonutChartLegend`, each wrapping a  `DonutChartLegendInfo`, with `DonutChartLegendInfo.value` being the total time elapsed for this project.
- [x] Create the actual cell/view, and populate it with the appropriate data. Designs are [here for iOS](https://toggl.invisionapp.com/d/main#/console/19800511/418226532/preview#project_console) and are still pending for Droid.
- [x] Discuss possible value animation with the team in #mobile-team

### Dependencies
Depends on #570

### PR DESCRIPTION
**Pull request by heytherewill**
### Summary
This adds a GLORIOUS compose UI to our reports. It makes me so happy

### Technical considerations
Compose is leaking from the settings module, which is equal parts exciting and frightening

### Why is it done like this and not some other way?
I implemented most of the UI, since it's really simple, but the canvas based ones will be done on another pull request for simplicity sake

### Relationships
Depends on #722

Closes #568
Closes #570
Closes #571
Closes #572
Closes #573
Closes #574
Closes #577

// More complex canvas based cards are to be implemented in a different Pull request to make reviewing easier
Refs #575
Refs #576

### Pull Request checklist
- [X] The code does not hinder the app's performance
- [ ] The code does not introduce regressions
- [ ] The code in this PR is tested thoroughly enough that changing the behavior to an incorrect one breaks tests

### Pointers
Go at it commit by commit

### Review template
- [ ] The code is thoroughly tested in well written and resilient manner
- [ ] The code is performant and leads to an optimal user experience
- [ ] The code is testable and does not cut corners for the sake of delivering
- [ ] The code considers older platforms well and will work in older devices just like it will in new ones
- [ ] The code considers light and dark theme
- [ ] The code considers RTL and does not mention left/right
- [ ] The code considers i18n
  - [ ] There are no hardcoded strings in the UI
  - [ ] All strings are included in every language we support OR there's an issue requesting new translations
- [ ] There are no hard-coded numbers referring to dimensions or font sizes
- [ ] The code considers UI elements that are similar and extracts their similarities into styles to avoid duplication

## Merge permissions
Only myself

### CODE REVIEW

File: reports/src/main/java/com/toggl/reports/domain/ReportsSelector.kt [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+            ),
+            ReportsViewModel.DonutChartLegend(
+                legend = DonutChartLegendInfo(
+                    "Mobile apps",
+                    "#F1C33F",
+                    "5%",
+                    "7:12:00"
+                )
+            ),
+            ReportsViewModel.Error,
```
Comment (semanticer) [inline reports/src/main/java/com/toggl/reports/domain/ReportsSelector.kt:46]: You should pass a `errorMessage` right?

File: reports/src/main/java/com/toggl/reports/ui/composables/AdvancedReportsOnWeb.kt [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+@Composable
+fun AdvancedReportsOnWeb(dispatcher: (ReportsAction) -> Unit) {
+
+    ConstraintLayout(Modifier.fillMaxWidth()) {
+
+        val (header, text, callToAction, image) = createRefs()
+
+        GroupHeader(
+            text = stringResource(R.string.advanced_reports),
+            modifier = Modifier.padding(top = grid_2) then Modifier.constrainAs(header) {
```
Comment (semanticer) [inline reports/src/main/java/com/toggl/reports/ui/composables/AdvancedReportsOnWeb.kt:31]: Why you don't simply chain these modifiers?
```suggestion
            modifier = Modifier.padding(top = grid_2).constrainAs(header) {
```

File: reports/src/main/java/com/toggl/reports/ui/composables/AdvancedReportsOnWeb.kt [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+                top.linkTo(parent.top)
+                start.linkTo(parent.start)
+            }
+        )
+
+        Text(
+            text = stringResource(R.string.advanced_reports_message),
+            style = MaterialTheme.typography.subtitle2,
+            modifier = Modifier.padding(top = grid_2) then Modifier.constrainAs(text) {
```
Comment (semanticer) [inline reports/src/main/java/com/toggl/reports/ui/composables/AdvancedReportsOnWeb.kt:40]: ⛓

File: reports/src/main/java/com/toggl/reports/ui/composables/AdvancedReportsOnWeb.kt [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+                top.linkTo(header.bottom)
+            }
+        )
+
+        Text(
+            text = stringResource(R.string.available_on_other_plans),
+            style = MaterialTheme.typography.subtitle2,
+            fontWeight = FontWeight.Bold,
+            modifier = Modifier.clickable(onClick = { dispatcher(ReportsAction.AvailableOnOtherPlansTapped) }) then
+                Modifier.padding(top = grid_2, bottom = grid_2) then Modifier.constrainAs(callToAction) {
```
Comment (semanticer) [inline reports/src/main/java/com/toggl/reports/ui/composables/AdvancedReportsOnWeb.kt:51]: ⛓

File: reports/src/main/java/com/toggl/reports/ui/composables/AdvancedReportsOnWeb.kt [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+                start.linkTo(parent.start)
+                top.linkTo(text.bottom)
+            }
+        )
+
+        Image(
+            asset = imageResource(id = R.drawable.illustration_graph),
+            modifier = Modifier.width(120.dp) then
+                Modifier.height(90.dp) then
```
Comment (semanticer) [inline reports/src/main/java/com/toggl/reports/ui/composables/AdvancedReportsOnWeb.kt:60]: ⛓

File: reports/src/main/java/com/toggl/reports/ui/composables/Billable.kt [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+                top.linkTo(parent.top)
+                start.linkTo(parent.start)
+            }
+        )
+
+        Text(
+            text = billableData.totalBillableTime,
+            style = MaterialTheme.typography.h4,
+            modifier = Modifier.padding(top = grid_2) then
+                Modifier.constrainAs(totals) {
```
Comment (semanticer) [inline reports/src/main/java/com/toggl/reports/ui/composables/Billable.kt:44]: ⛓

File: reports/src/main/java/com/toggl/reports/ui/composables/Billable.kt [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+                    start.linkTo(parent.start)
+                    top.linkTo(header.bottom)
+                }
+        )
+
+        Text(
+            text = "${billableData.billablePercentage}%",
+            style = MaterialTheme.typography.subtitle2,
+            modifier = Modifier.padding(top = grid_3) then
+                Modifier.constrainAs(percent) {
```
Comment (semanticer) [inline reports/src/main/java/com/toggl/reports/ui/composables/Billable.kt:54]: ⛓

File: reports/src/main/java/com/toggl/reports/ui/composables/Billable.kt [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+                    bottom.linkTo(bar.top)
+                }
+        )
+
+        Box(
+            backgroundColor = Color(0xFF999999),
+            modifier = Modifier.padding(top = grid_0_5) then
+                Modifier.height(2.dp) then
+                Modifier.width(billableChartWidth.dp) then
+                Modifier.constrainAs(bar) {
```
Comment (semanticer) [inline reports/src/main/java/com/toggl/reports/ui/composables/Billable.kt:66]: ⛓

File: reports/src/main/java/com/toggl/reports/ui/composables/Billable.kt [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+                }
+        )
+
+        val filledWidth = (billableChartWidth / 100F) * billableData.billablePercentage
+
+        Box(
+            backgroundColor = MaterialTheme.colors.primary,
+            modifier = Modifier.height(4.dp) then
+                Modifier.width(filledWidth.dp) then
+                Modifier.constrainAs(filledBar) {
```
Comment (semanticer) [inline reports/src/main/java/com/toggl/reports/ui/composables/Billable.kt:78]: ⛓

File: reports/src/main/java/com/toggl/reports/ui/composables/Common.kt [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+        modifier = Modifier.padding(bottom = grid_2) then modifier
+    )
+}
+
+@Composable
+internal fun SectionDivider(
+    modifier: Modifier = Modifier
+) {
+    Divider(
+        modifier = modifier then Modifier.padding(
```
Comment (semanticer) [inline reports/src/main/java/com/toggl/reports/ui/composables/Common.kt:44]: ⛓

File: reports/src/main/java/com/toggl/reports/ui/composables/DonutChartLegend.kt [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+        Box(
+            shape = CircleShape,
+            backgroundColor = projectColor,
+            modifier = Modifier.size(10.dp)
+        )
+
+        Text(
+            text = legend.projectName,
+            modifier = Modifier.weight(8F, fill = true)
+                .then(Modifier.padding(start = grid_2))
```
Comment (semanticer) [inline reports/src/main/java/com/toggl/reports/ui/composables/DonutChartLegend.kt:42]: ⛓

File: settings/src/main/java/com/toggl/settings/ui/common/ChoiceListWithHeader.kt [hunk 12]
```diff
@@ -75,12 +78,14 @@ internal fun SingleChoiceListWithHeader(
         closeAction?.run(dispatcher)
     }
 
-    Column(modifier = Modifier.padding(grid_3) + Modifier.fillMaxWidth()) {
+    Column(modifier = Modifier.padding(grid_3) then Modifier.fillMaxWidth()) {
```
Comment (semanticer) [inline settings/src/main/java/com/toggl/settings/ui/common/ChoiceListWithHeader.kt:81]: ⛓

File: settings/src/main/java/com/toggl/settings/ui/common/ChoiceListWithHeader.kt [hunk 13]
```diff
@@ -94,8 +99,8 @@ internal fun SingleChoiceListWithHeader(
             items.forEach { item ->
                 Box(modifier = Modifier.clickable(onClick = { dispatchOnSelected(item) })) {
                     Row(
-                        modifier = Modifier.preferredHeight(grid_6) +
-                            Modifier.clickable(onClick = { dispatchOnSelected(item) }) +
+                        modifier = Modifier.preferredHeight(grid_6) then
```
Comment (semanticer) [inline settings/src/main/java/com/toggl/settings/ui/common/ChoiceListWithHeader.kt:102]: ⛓

File: settings/src/main/java/com/toggl/settings/ui/common/Section.kt [hunk 14]
```diff
@@ -87,9 +85,7 @@ internal fun SettingsRow(
             )
         else Modifier
 
-    val rowModifier = Modifier
-        .height(48.dp)
-        .fillMaxWidth() + clickModifier
+    val rowModifier = Modifier.height(48.dp).fillMaxWidth() then clickModifier
```
Comment (semanticer) [inline settings/src/main/java/com/toggl/settings/ui/common/Section.kt:88]: This is actually the right use case for `then`

File: reports/src/main/java/com/toggl/reports/ui/composables/Amounts.kt [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+                Text(
+                    text = amount.currency,
+                    style = MaterialTheme.typography.subtitle1,
+                    fontWeight = FontWeight.Medium
+                )
+            }
+        }
+
+        SectionDivider()
+    }
```
Comment (semanticer) [inline reports/src/main/java/com/toggl/reports/ui/composables/Amounts.kt:46]: I might be overthinking this but it looks like you are repeating 

```kotlin
Column {
   ?GroupHeader()
   SectionHeader()
   BlockOfContent()
   SectionDivider()
}
```
You could theoretically turn this into custom Composable and use it like this:

```kotlin
ReportSection(
  groupHeader = "title",
  sectionHeader = "title"
) {
   BlockOfContent()
}
```
But feel free to ignore this suggestion

**Comment by heytherewill** [comment]
> Donut chart texts are "quite" small on my hi-res emulator.

This does not close the donut chart issue; we will do it properly somewhere else

> It crashes when I tap on "available on other plans" but I guess that's going to be implemented later as well.

Yeah, there are no issues for adding this view to aurora yet

**Pull request merged**